lib/models/per-bundle-addon-cache/index.js

```diff
@@ -0,0 +1,402 @@
+          // Fill in the entry and return the new instance.
+          logger.debug(`About to fill in REGULAR ADDON EXISTING cache entry for addon at: ${addonPkgInfo.realPath}`);
+          this.numAddonInstances++;
+          addonInstance = addonPkgInfo.constructAddonInstance(parent, this.project);
+          addonCacheEntry[TARGET_INSTANCE] = addonInstance; // cache BEFORE initializing child addons
+          addonPkgInfo.initChildAddons(addonInstance);
+          return addonInstance;
+        }
+      }
+
+      // There is no entry for this addon in the bundleHost's addon cache. Create a new
+      // instance, cache it in the addon cache, and return it.
+      logger.debug(`About to construct REGULAR ADDON NEW cache entry for addon at: ${addonPkgInfo.realPath}`);
+      this.numAddonInstances++;
+      addonInstance = addonPkgInfo.constructAddonInstance(parent, this.project);
+      addonCacheEntry = this.createAddonCacheEntry(addonInstance, addonPkgInfo.realPath);
+      addonInstanceCache.set(addonPkgInfo.realPath, addonCacheEntry); // cache BEFORE initializing child addons
+      addonPkgInfo.initChildAddons(addonInstance);
+      return addonInstance;
+    } else {
+      // The bundleHost is not the project but the project bundles the addon too and
+      // deduping is enabled, so the cache entry needs to go in the project's cache.
+      // Get/create an empty cache entry and return a proxy to it. The project will
+      // set the instance later (see above).
+      let bundleHostCacheEntry = this.bundleHostCache.get(PROJECT_BUNDLE_HOST_NAME);
+      let addonCacheEntry = bundleHostCacheEntry.addonInstanceCache.get(addonPkgInfo.realPath);
+      if (!addonCacheEntry) {
+        logger.debug(`About to construct REGULAR ADDON EMPTY cache entry for addon at: ${addonPkgInfo.realPath}`);
+        addonCacheEntry = this.createAddonCacheEntry(null, addonPkgInfo.realPath);
+        bundleHostCacheEntry.addonInstanceCache.set(addonPkgInfo.realPath, addonCacheEntry);
+      }
+
+      logger.debug(`About to construct REGULAR ADDON PROXY for EMPTY addon at: ${addonPkgInfo.realPath}`);
+      this.numProxies++;
+      return getAddonProxy(addonCacheEntry, parent);
+    }
+  }
+
+  /**
+   * Called by engine-addon when it needs to retrieve the real child addons for a given
+   * object (i.e., during ancestorHostAddons), even though the object might be a proxy.
+   *
+   * @method getAddonAddons
+   * @param {Addon|Proxy} addonInstance an instance of an addon or a Proxy generated by this class
+   * @return {Array} an array of addon instances
+   */
+  getAddonAddons(addonInstance) {
+    let result =
+      addonInstance[TARGET_INSTANCE] && addonInstance[TARGET_INSTANCE].addons
+        ? addonInstance[TARGET_INSTANCE].addons
+        : addonInstance.addons;
+    return result;
+  }
+}
+
+module.exports = PerBundleAddonCache;
+
+// Support for per-bundle addon caching is opt OUT. Caching any particular
```

Comment:
i'd think i'd clarify this a bit; it's a _global_ opt-out (even if any addon has itself enabled `allowCachingPerBundle`)

Reply:
Sounds good.

Reply:
Updated the comment to be more explanatory.